### PR TITLE
Improved customwalls/regular walls collision and player push

### PR DESCRIPTION
**Pull request by IvoryDuke**
Reworked CPlayer::push() for customwalls to properly push player out of danger if that is possible.
Tests on levels making use of customwalls show expected results.
By default all customwalls are an instakill but they can be set to behave like regular walls and even be more lenient with two new flags.
cw_setDeadly true/false. If false they behave similarly to regular walls.
cw_setForgiving true/false. If true they are more lenient toward saving player. This is a required settings if level maker want player to slide along particularly exotic walls.

Suggested levels for testing:
- Experimental "Arcadia" (custom walls);
- Experimental "Curvetest" (curving walls);
- Experimental "Collision Test" (pulsing levels);
- Polygonal Cubics "Oh No!" (extreme customwalls);
- Polygonal Cubics "How?!?" (extreme customwalls);
- Polygonal Cubics "Misdirection" (customwalls on level with really high amount of sides);
- Attached level zip (standard walls on level with really high amount of sides).

[sandbox.zip](https://github.com/SuperV1234/SSVOpenHexagon/files/5962613/sandbox.zip)

### CODE REVIEW

**Comment by IvoryDuke** [comment]
I ran into an unexpected issue which will require a little more time to figure out.
I am very close to a viable solution but need to address a rather annoying bug before I update the PR.

File: _RELEASE/Assets/favoriteLevels.json [hunk 1]
```diff
@@ -2,10 +2,7 @@
 {
 	"ids" : 
 	[
-		"LoFiOHv1_Jocii_Chill_Open_Hexagon_to_Calm_the_Vibe_1_LoFi01",
-		"ohvrvanilla_vittorio_romeo_cube_1_apeirogon",
-		"ohvrvanilla_vittorio_romeo_cube_1_commando",
-		"ohvrvanilla_vittorio_romeo_experimental_1_curvetest",
-		"ohvrvanilla_vittorio_romeo_experimental_1_curvetest2"
+		"ohvrvanilla_vittorio_romeo_experimental_1_arcadia",
```
Comment (SuperV1234) [inline _RELEASE/Assets/favoriteLevels.json:5]: Favorite levels should be saved on a per-profile basis, but I'm fine with that being done for a future update. Also this file should be cleared up before merging.

File: include/SSVOpenHexagon/Components/CPlayer.hpp [hunk 2]
```diff
@@ -21,9 +23,13 @@ class CCustomWall;
...
 {
 private:
-    sf::Vector2f startPos;
-    sf::Vector2f pos;
-    sf::Vector2f lastPos;
+    sf::Vector2f startPos;      // Position at start of the level.
+    sf::Vector2f pos;           // Actual position of player.
+    sf::Vector2f prePushPos;    // Position before player is pushed by a wall.
+                                // Unline pos it is not updated after a successful
+                                // wall push.
```
Comment (SuperV1234) [inline include/SSVOpenHexagon/Components/CPlayer.hpp:30]:
```suggestion
    sf::Vector2f prePushPos;    // Position before the player is pushed by a wall.
                                // Unlike `pos` it is not updated after a successful
                                // wall push.
```

File: include/SSVOpenHexagon/Components/CPlayer.hpp [hunk 3]
```diff
@@ -21,9 +23,13 @@ class CCustomWall;
...
+    sf::Vector2f lastPos;       // Position of player in the previous frame, adjusted
+                                // according to the current frame's radius.
```
Comment (SuperV1234) [inline include/SSVOpenHexagon/Components/CPlayer.hpp:32]:
```suggestion
    sf::Vector2f lastPos;       // Position of the player in the previous frame, adjusted
                                // according to the current frame's radius.
```

File: include/SSVOpenHexagon/Components/CPlayer.hpp [hunk 4]
```diff
@@ -34,6 +40,11 @@ class CPlayer
 
     bool dead;
     bool justSwapped;
+    bool forcedMove;
+
+    float radius;
+    float maxSafeDistance;
+    float currentSpeed;
```
Comment (SuperV1234) [inline include/SSVOpenHexagon/Components/CPlayer.hpp:47]: These fields should be documented as well, otherwise it will be hard to keep track of everything a few weeks from now :)

File: include/SSVOpenHexagon/Components/CWall.hpp [hunk 5]
```diff
@@ -28,31 +28,32 @@ class CWall
     float hueMod;
     bool killed;
 
-public:
-    CWall(HexagonGame& mHexagonGame, const sf::Vector2f& mCenterPos, int mSide,
-        float mThickness, float mDistance, const SpeedData& mSpeed,
-        const SpeedData& mCurve);
-
-    void update(const HexagonGame& mHexagonGame, const ssvu::FT mFT);
+    bool outOfPlayerRadius;
```
Comment (SuperV1234) [inline include/SSVOpenHexagon/Components/CWall.hpp:31]: Ditto, please document

File: include/SSVOpenHexagon/Components/CWall.hpp [hunk 6]
```diff
@@ -67,12 +68,40 @@ class CWall
...
+        // If the wall is too far from the center is cannot cause collision.
+        // If it is and both distance and pointInPolygon() calculations are executed
+        // we lose performance, If not (which is the case in the vast majority of
+        // any level's runtime) we only do a much faster algorithm.
+        if(ssvs::getMagSquared((vertexPositions[0] + vertexPositions[1]) / 2.f -
+            mCenterPos) > mRadiusSquared)
+        {
+            outOfPlayerRadius = true;
+            return false;
+        }
```
Comment (SuperV1234) [inline include/SSVOpenHexagon/Components/CWall.hpp:84]: This seems like a nice little optimization. Have you actually measured the impact? Would be interesting to know. 
Also wondering if we can avoid the call to `getMagSquared` somehow and avoid the extra multiplication, maybe we could cache that value once per frame?

File: include/SSVOpenHexagon/Utils/PointInPolygon.hpp [hunk 7]
```diff
@@ -12,12 +12,12 @@ template <typename TC, typename T>
     const TC& mVertices, T x, T y) noexcept
 {
     bool result{false};
+    auto size{mVertices.size()};
```
Comment (SuperV1234) [inline include/SSVOpenHexagon/Utils/PointInPolygon.hpp:15]:
```suggestion
    const auto size{mVertices.size()};
```

File: include/SSVOpenHexagon/Utils/PointInPolygon.hpp [hunk 8]
```diff
@@ -29,4 +29,26 @@ template <typename TC, typename T>
     return result;
 }
 
+template <typename TC, typename T>
+[[gnu::always_inline, gnu::pure, nodiscard]] inline bool pointInPolygonPointers(
+    const TC& mVertices, T x, T y) noexcept
+{
+    bool result{false};
+    auto size{mVertices.size()};
```
Comment (SuperV1234) [inline include/SSVOpenHexagon/Utils/PointInPolygon.hpp:37]:
```suggestion
    const auto size{mVertices.size()};
```

File: include/SSVOpenHexagon/Utils/Utils.hpp [hunk 9]
```diff
@@ -43,19 +43,39 @@ inline void uppercasify(std::string& s)
     return s;
 }
 
-inline float getFontHeight(sf::Text& font)
+[[nodiscard, gnu::pure]] inline float getFontHeight(sf::Text& font)
```
Comment (SuperV1234) [inline include/SSVOpenHexagon/Utils/Utils.hpp:46]:
```suggestion
[[nodiscard]] inline float getFontHeight(sf::Text& font)
```

`gnu::pure` can be a bit problematic with references involved. I am not sure it can be applied correctly here, so please remove it.

File: include/SSVOpenHexagon/Utils/Utils.hpp [hunk 10]
```diff
@@ -43,19 +43,39 @@ inline void uppercasify(std::string& s)
...
 {
     font.setString("A");
     return ssvs::getGlobalHeight(font);
 }
 
-inline float getFontHeight(sf::Text& font, const unsigned int charSize)
+[[nodiscard, gnu::pure]] inline float getFontHeight(sf::Text& font, const unsigned int charSize)
```
Comment (SuperV1234) [inline include/SSVOpenHexagon/Utils/Utils.hpp:52]:
```suggestion
[[nodiscard]] inline float getFontHeight(sf::Text& font, const unsigned int charSize)
```

Ditto

File: include/SSVOpenHexagon/Utils/Utils.hpp [hunk 11]
```diff
@@ -43,19 +43,39 @@ inline void uppercasify(std::string& s)
...
+    const sf::Vector2f& p2, const float mRadiusSquared);
+
+bool getLineCircleClosestIntersection(
+    sf::Vector2f& mIntersection, const sf::Vector2f& mPos,
+    const sf::Vector2f& p1, const sf::Vector2f& p2,
+    const float mRadiusSquared);
+
+[[nodiscard, gnu::pure]] inline float fastSqrt(const float &n) 
+{
+    static union{int i; float f;} u;
```
Comment (SuperV1234) [inline include/SSVOpenHexagon/Utils/Utils.hpp:74]:
```suggestion
    union{int i; float f;} u;
```

Why `static`?

File: include/SSVOpenHexagon/Utils/Utils.hpp [hunk 12]
```diff
@@ -43,19 +43,39 @@ inline void uppercasify(std::string& s)
...
+unsigned int getLineCircleIntersection(sf::Vector2f& i1,
+    sf::Vector2f& i2, const sf::Vector2f& p1,
+    const sf::Vector2f& p2, const float mRadiusSquared);
+
+bool getLineCircleClosestIntersection(
+    sf::Vector2f& mIntersection, const sf::Vector2f& mPos,
+    const sf::Vector2f& p1, const sf::Vector2f& p2,
+    const float mRadiusSquared);
+
+[[nodiscard, gnu::pure]] inline float fastSqrt(const float &n)
```
Comment (SuperV1234) [inline include/SSVOpenHexagon/Utils/Utils.hpp:72]: Did you measure that this is actually faster than `std::sqrt`?

File: src/SSVOpenHexagon/Components/CPlayer.cpp [hunk 13]
```diff
@@ -159,112 +152,242 @@ void CPlayer::kill(HexagonGame& mHexagonGame)
...
-    if(!getJustSwapped())
+template <typename Wall>
+[[nodiscard]] bool CPlayer::checkWallCollisionEscape(const Wall& wall,
+    sf::Vector2f& mPos, const float mRadiusSquared)
+{
+    // To find the closest wall side we intersect the circumference of the possible
+    // player positions with the sides of the wall. We use the intersection closest
+    // to the player's position as post collision target.
+    // If an escape route could not be found player is killed.
+
```
Comment (SuperV1234) [inline src/SSVOpenHexagon/Components/CPlayer.cpp:167]: This is all pretty arcane for me but I'm just gonna trust you :)

File: src/SSVOpenHexagon/Components/CPlayer.cpp [hunk 14]
```diff
@@ -159,112 +152,242 @@ void CPlayer::kill(HexagonGame& mHexagonGame)
...
-    const sf::Vector2f testPos{
-        ssvs::getOrbitRad(startPos, testAngle, mHexagonGame.getRadius())};
-
-    // If there is overlap even after compensation kill without updating
-    // position, as there is no benefit in doing it.
-    if(wall.isOverlapping(testPos))
+    // ...otherwise make testPos the position of the previous frame plus
+    // the curving wall's velocity, and check an escape on that position.
+    // Using the previous frame's position is essential for levels with
+    // a really high amount of sides.
```
Comment (SuperV1234) [inline src/SSVOpenHexagon/Components/CPlayer.cpp:247]: Why? Would be nice to know in the comment

File: src/SSVOpenHexagon/Components/CPlayer.cpp [hunk 15]
```diff
@@ -159,112 +152,242 @@ void CPlayer::kill(HexagonGame& mHexagonGame)
...
+    if(wall.isOverlapping(testPos) ||
+       !checkWallCollisionEscape(wall, testPos, mRadiusSquared))
+    {
+        return true;
+    }
+
+    // If player survived apply test position and add a little padding.
+    pos = testPos + ssvs::getNormalized(testPos - prePushPos) * collisionPadding;
+    angle = ssvs::getRad(pos);
+    updatePosition(mHexagonGame, mFT);
```
Comment (SuperV1234) [inline src/SSVOpenHexagon/Components/CPlayer.cpp:390]: Ditto, this is also pretty arcane to me (the maths), but I'm gonna trust (and test)

File: src/SSVOpenHexagon/Components/CWall.cpp [hunk 16]
```diff
@@ -79,17 +85,15 @@ void CWall::moveTowardsCenter(HexagonGame& mHexagonGame,
         ssvs::moveTowards(vp, mCenterPos, speed.speed * 5.f * mFT);
     }
 
-    if(pointsOnCenter > 3 || pointsOutOfBounds > 3)
+    if(pointsOnCenter == 4 || pointsOutOfBounds == 4)
```
Comment (SuperV1234) [inline src/SSVOpenHexagon/Components/CWall.cpp:88]: Why was this changed? Maybe there could be a weird scenario where a weirdly shaped wall has its vertices counted multiple times... or maybe it can never happen...?

**Review by SuperV1234 — CHANGES_REQUESTED** [review]
Looks generally good! Will have to test a lot and make sure everything works, but this is awesome. Just a few things to improve/fix

File: include/SSVOpenHexagon/Utils/Utils.hpp (hunk 12 above)
Comment (IvoryDuke) [reply include/SSVOpenHexagon/Utils/Utils.hpp:72]: It is a well known algorithm that has been used since the Quake engine.
I have not measured myself how it compares to std::sqrt, I will inquire when possible.

File: src/SSVOpenHexagon/Components/CWall.cpp (hunk 16 above)
Comment (IvoryDuke) [reply src/SSVOpenHexagon/Components/CWall.cpp:88]: I made it == 4 cause I always found it to be somewhat misleading to read > 3, which to me implied there could be polygons with an arbitrary number of vertices higher than 3.
I do not mind reverting it though.

File: include/SSVOpenHexagon/Components/CWall.hpp (hunk 6 above)
Comment (IvoryDuke) [reply include/SSVOpenHexagon/Components/CWall.hpp:84]: I am not sure how we could avoid using getMagSquared(), I am already skipping the unnecessary step of running sqrt on it since if x > y => x * x > y * y but that is as far as I think we can go without running in inaccurate calculations.
Caching would provide no benefits since that specific overload is called only the first collision check of a frame, the second one relies on
```
    [[gnu::always_inline, nodiscard]] bool isOverlapping(
        const sf::Vector2f& mPoint) const noexcept
    {
        if(outOfPlayerRadius)
        {
            return false;
        }

        return Utils::pointInPolygon(vertexPositions, mPoint.x, mPoint.y);
    }
```
I will see to measure the impact one of the next days.

File: include/SSVOpenHexagon/Utils/Utils.hpp (hunk 11 above)
Comment (IvoryDuke) [reply include/SSVOpenHexagon/Utils/Utils.hpp:74]: I copied-pasted the algorithm from a game development website. Gonna make it static.

File: include/SSVOpenHexagon/Utils/Utils.hpp (hunk 12 above)
Comment (IvoryDuke) [reply include/SSVOpenHexagon/Utils/Utils.hpp:72]: Run a simple test, this example code returns 0 for the duration of fastSqrt and 1 for std::sqrt
```
#include <iostream>
#include <chrono>
#include <cmath>

inline float fastSqrt(const float &n) 
{
    union{int i; float f;} u;
    u.i = 0x5F375A86 - (*(int*)&n >> 1);
    return (int(3) - n * u.f * u.f) * n * u.f * 0.5f;
}

int main()
{
    const float num = 38725673265436589345.324677;
    
    auto t1 = std::chrono::high_resolution_clock::now();
    const auto f = fastSqrt(num);
    auto t2 = std::chrono::high_resolution_clock::now();
    std::cout << std::chrono::duration_cast<std::chrono::microseconds>( t2 - t1 ).count() << std::endl;
    
    auto t3 = std::chrono::high_resolution_clock::now();
    const auto g = std::sqrt(num);
    auto t4 = std::chrono::high_resolution_clock::now();
    std::cout << std::chrono::duration_cast<std::chrono::microseconds>( t4 - t3 ).count() << std::endl;
    
    return 0;
}
```

**Comment by IvoryDuke** [comment]
I have added explanatory comments on all requested parts of the code and have applied suggested tweaks.

File: include/SSVOpenHexagon/Components/CWall.hpp (hunk 6 above)
Comment (IvoryDuke) [reply include/SSVOpenHexagon/Components/CWall.hpp:89]: Example code returns 2 0
` ` ` 
#include <iostream>
#include <chrono>
#include <array>

struct Point
{
    float x;
    float y;
};

inline bool pointInPolygon(const std::array<Point, 4>& points, float x, float y) noexcept
{
    bool result{false};

    for(unsigned int i{0}, j{3}; i < 4; j = i++)
    {
        const auto& vI{points[i]};
        const auto& vJ{points[j]};

        if(((vI.y > y) != (vJ.y > y)) &&
            (x < (vJ.x - vI.x) * (y - vI.y) / (vJ.y - vI.y) + vI.x))
        {
            result = !result;
        }
    }

    return result;
}

inline float magSquared(Point& p) noexcept
{
    return p.x * p.x + p.y * p.y;
}

int main()
{
    auto t1 = std::chrono::high_resolution_clock::now();
    std::array<Point, 4> points = {Point{0.f, 0.f}, Point{5.f, 0.f}, Point{10.f, 22.f}, Point{-2.f, 6.f}};
    const auto b = pointInPolygon(points, 4.f, 6.f);
    auto t2 = std::chrono::high_resolution_clock::now();
    std::cout << std::chrono::duration_cast<std::chrono::microseconds>( t2 - t1 ).count() << std::endl;

    auto t3 = std::chrono::high_resolution_clock::now();
    Point p = {20.f, 16.f};
    const auto f = magSquared(p);
    auto t4 = std::chrono::high_resolution_clock::now();
    std::cout << std::chrono::duration_cast<std::chrono::microseconds>( t4 - t3 ).count() << std::endl;

    return 0;
}
` ` `

**Comment by IvoryDuke** [comment]
Replaced forgiving status with the possibility for users to choose which side should definitively kill player.
This both solves a few rare unjustified deaths and allows level makers to make more precise and deliberate choices on the behavior of the walls.
The reason why collision on standard walls is precise and reliable is that we know for sure that the only side of it that should kill the player is the one closer to the center (the one delimited by vertexes 0 and 1). This puts more work on level makers but uniforms behavior between custom and regular walls when used correctly.

**Comment by SuperV1234** [comment]
> Replaced forgiving status with the possibility for users to choose which side should definitively kill player.

Do you think you can finalize the work on custom walls soon? Or maybe we should release it as experimental? It's been a while since a new version release :)

**Comment by IvoryDuke** [comment]
I actually consider this to be final, the thing I just did randomly popped into my mind today.
It was very functional even before, but this makes it even more reliable.

**Review by SuperV1234 — APPROVED** [review]
